src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,412 @@
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using TransformedTargetType = typename Microsoft::Featurizer::Traits<InputT>::nullable_type;
+
+    using BaseType =
+        Transformer<
+            InputT,
+            Microsoft::Featurizer::RowMajMatrix<TransformedTargetType>
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets);
+    LagLeadOperatorTransformer(Archive &ar);
+
+    ~LagLeadOperatorTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(LagLeadOperatorTransformer);
+
+    bool operator==(LagLeadOperatorTransformer const &other) const;
+    bool operator!=(LagLeadOperatorTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    std::uint32_t const                                _horizon;
+    std::int64_t const                                 _offset_delta;
+    std::vector<std::int64_t> const                    _offsets;
+    Components::CircularBuffer<TransformedTargetType>  _buffer;
+    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        execute_helper(TransformedTargetType(input), callback);
+    }
+
+    void flush_impl(typename BaseType::CallbackFunction const & callback) override {
+        std::int64_t numPending = 0;
+        std::int64_t max_future((_offset_delta + *std::max_element(_offsets.cbegin(), _offsets.cend())) > 0 ? (_offset_delta + *std::max_element(_offsets.cbegin(), _offsets.cend())) : 0);
+
+        while (numPending < max_future) {
+            execute_helper(Traits<TransformedTargetType>::CreateNullValue(), callback);
+            ++numPending;
+        }
+        // after flush is called, this transfomer should be reset to its starting state
+        _buffer.clear();
+        for (std::int64_t i = 0; i < (static_cast<std::int64_t>(_horizon) - _offset_delta - 1); ++i) {
+            _buffer.push(Traits<TransformedTargetType>::CreateNullValue());
+        }
```

Comment:
This code is duplicated in the ctor as well. Implement it in a single place.

Reply:
Created a separate function to prepopulate nulls.